### PR TITLE
chore(gatsby-page-utils): Migrate package to TypeScript

### PR DESCRIPTION
**Pull request by arthurjdam**
## Description

Migrates `gatsby-page-utils` to Typescript

## Related Issues

Related to #21995

### CODE REVIEW

**Comment by blainekasten** [comment]
Thanks for your hard work on this! Give me a few days to circle back to this one to review. The community support on this migration has been overwhelming! It's wonderful 💜

**Comment by LekoArts** [comment]
Sorry for the late response on this. Can you please rebase onto master? Thanks!

File: packages/gatsby-page-utils/src/ignore-path.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
+import * as mm from "micromatch"
+
+interface IPathIgnoreOptions {
+  patterns: string | ReadonlyArray<string>
+  options?: mm.Options
+}
+
+export function ignorePath(
+  path: string,
+  ignore?: IPathIgnoreOptions | string | string[] | {} | null
```
Comment (pvdz) [inline packages/gatsby-page-utils/src/ignore-path.ts:10]: If you change `IPathIgnoreOptions` to have an optional `pattern` field, then you can probably just drop `{}` from the type of `ignore` here.

File: packages/gatsby-page-utils/src/ignore-path.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  } else {
+    // Allow shorthand ignore patterns ['pattern'] or 'pattern'
+    if (Array.isArray(ignore) && ignore.length > 0) {
+      settings.patterns = ignore
+    } else if (typeof ignore === `string`) {
+      settings.patterns = ignore
+    } else {
+      return false
+    }
+  }
```
Comment (pvdz) [inline packages/gatsby-page-utils/src/ignore-path.ts:35]: If you reorder the checks then I expect TS to be able to infer the proper type so you don't have to cast.

```suggestion
  if (typeof ignore === `string`) {
    settings.patterns = ignore
  } else if (Array.isArray(ignore)) {
    if (ignore.length > 0) {
      settings.patterns = ignore
    }
  } else if (ignore === null) {
    return false
  } else {
    // TS should be able to assert that `ignore` must be a `IPathIgnoreOptions` now
    if (!ignore.options && !ignore.patterns) {
      return false
    }
    if (ignore.options) {
      settings.options = ignore.options
    }
    if (ignore.patterns) {
      settings.patterns = ignore.patterns
    }
  }    
```

File: packages/gatsby-page-utils/src/ignore-path.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  // Return true if the path should be ignored (matches any given ignore patterns)
+  return mm.any(path, settings.patterns as string | string[], settings.options)
```
Comment (pvdz) [inline packages/gatsby-page-utils/src/ignore-path.ts:37]: With the above I'd half expect the casting in this line not to be required

File: packages/gatsby-page-utils/src/watch-directory.ts [hunk 4]
```diff
@@ -1,9 +1,13 @@
-const Promise = require(`bluebird`)
```
Comment (pvdz) [inline packages/gatsby-page-utils/src/watch-directory.ts:1]: Dropping bluebird could be a non-trivial change. Did you actually check whether this holds or was this more of a cleanup thing? While I'm in favor of removing bluebird where possible, in favor of native promises, these changes should not be part of the TS migration.

Please undo unless you can guarantee the change is trivial.

File: packages/gatsby-page-utils/src/watch-directory.ts [hunk 5]
```diff
@@ -1,9 +1,13 @@
...
-const chokidar = require(`chokidar`)
+import * as chokidar from "chokidar"
 const { slash } = require(`gatsby-core-utils`)
 
-module.exports = async (path, glob, onNewFile, onRemovedFile) =>
-  new Promise((resolve, reject) => {
+export function watchDirectory(
```
Comment (pvdz) [inline packages/gatsby-page-utils/src/watch-directory.ts:4]: Why did you drop the `async`? Please restore that keyword. It won't change the runtime but it helps the tooling and the reviewers.

File: packages/gatsby-page-utils/package.json [hunk 6]
```diff
@@ -27,11 +27,12 @@
     "gatsby-core-utils": "^1.3.5",
     "glob": "^7.1.6",
     "lodash": "^4.17.15",
-    "micromatch": "^3.1.10"
```
Comment (pvdz) [inline packages/gatsby-page-utils/package.json:30]: Did you intentionally major-bump micromatch? I've seen a few reported issues related to micromatch bumping recently so I'd prefer not to change this if at all possible.

File: packages/gatsby-page-utils/package.json [hunk 7]
```diff
@@ -4,8 +4,8 @@
   "description": "Gatsby library that helps creating pages",
   "main": "dist/index.js",
   "scripts": {
-    "build": "babel src --out-dir dist/ --ignore \"**/__tests__\"",
-    "watch": "babel -w src --out-dir dist/ --ignore \"**/__tests__\"",
+    "build": "babel src --out-dir dist/ --ignore **/__tests__ --extensions \".ts,.js\"",
+    "watch": "babel -w src --out-dir dist/ --ignore **/__tests__ --extensions \".ts,.js\"",
```
Comment (pvdz) [inline packages/gatsby-page-utils/package.json:8]: Shouldn't the `**/__tests__` parts be double quoted as well? (They were before)

I think this makes the glob expansion safer in certain cases.

Comment (arthurjdam) [reply packages/gatsby-page-utils/package.json]: Fair. These shouldn't have been lost

File: packages/gatsby-page-utils/package.json (hunk 6 above)
Comment (arthurjdam) [reply packages/gatsby-page-utils/package.json]: There is a problem with the types for 3.1.10. 
Updated so it only uses the types from 4.0.1 now

File: packages/gatsby-page-utils/src/watch-directory.ts (hunk 5 above)
Comment (arthurjdam) [reply packages/gatsby-page-utils/src/watch-directory.ts]: Good catch

File: packages/gatsby-page-utils/src/ignore-path.ts (hunk 1 above)
Comment (arthurjdam) [reply packages/gatsby-page-utils/src/ignore-path.ts]: Done

File: packages/gatsby-page-utils/package.json (hunk 6 above)
Comment (pvdz) [reply packages/gatsby-page-utils/package.json]: Yeah that's super. Thanks! :)

File: packages/gatsby-page-utils/package.json (hunk 7 above)
Comment (pvdz) [reply packages/gatsby-page-utils/package.json]: I suspect you fixed this and then the fix got lost after the @types update? :)

**Comment by pvdz** [comment]
Thanks for the making the changes :) Looks really good!

I'm looking into the type error that CI is currently throwing because I don't expect this PR to cause it (but as it's a TS migration PR, who knows :shrug: ).

**Comment by pvdz** [comment]
Ok wow what happened to the yarn lock? :open_mouth: 

Could you try to drop changes to that file, run `yarn`, and see whether that fixes it?

**Comment by arthurjdam** [comment]
Yikes indeed! Must have had a wrong node version. The diff looks a lot more reasonable now

**Comment by pvdz** [comment]
(merged master in to work around known test breakage that was fixed last week).
If tests hold this can be merged, yay :D

**Comment by arthurjdam** [comment]
Thanks!

**Pull request merged**